### PR TITLE
New Rys kernel

### PR DESCRIPTION
**Pull request by sunqm**
* Optimize the Rys quadrature kernel for 4-center ERIs. New kernel is ~30% faster for DZ basis sets, and ~2 times faster for TZ basis sets than the old kernel.
* Add SR ERIs.
* Add J engine based on Rys integral kernel. J-engine kernel is >5 times faster than old kernel.
* Add a cupy_helper function "condense" corresponding to the "lib.condense" function in pyscf

### CODE REVIEW

File: gpu4pyscf/scf/jk.py [hunk 1]
```diff
@@ -0,0 +1,657 @@
...
+    if omega is None:
+        omega = mol.omega
+
+    mol = vhfopt.mol
+    nao, nao_orig = vhfopt.coeff.shape
+
+    dm = cp.asarray(dm, order='C')
+    dms = dm.reshape(-1,nao_orig,nao_orig)
+    n_dm = dms.shape[0]
+    dms = cp.einsum('pi,nij,qj->npq', vhfopt.coeff, dms, vhfopt.coeff)
```
Comment (wxj6000) [inline gpu4pyscf/scf/jk.py:60]: `cp.einsum` will consume a large amount of GPU memory for 3-tensor contraction

File: gpu4pyscf/scf/jk.py [hunk 2]
```diff
@@ -0,0 +1,657 @@
...
+                        timing_collection[llll] += t1[1] - t1p[1]
+                        kern_counts += 1
+
+    if log.verbose >= logger.DEBUG1:
+        log.debug1('kernel launches %d', kern_counts)
+        for llll, t in timing_collection.items():
+            log.debug1('%s wall time %.2f', llll, t)
+
+    if with_k:
+        vk = cp.einsum('pi,npq,qj->nij', vhfopt.coeff, vk, vhfopt.coeff)
```
Comment (wxj6000) [inline gpu4pyscf/scf/jk.py:159]: decompose `cp.einsum` into two tensor contractions

File: gpu4pyscf/scf/jk.py [hunk 3]
```diff
@@ -0,0 +1,657 @@
...
+        vk = vk + vk.transpose(0,2,1)
+        vk = vk.reshape(dm.shape)
+    if with_j:
+        vj = cp.einsum('pi,npq,qj->nij', vhfopt.coeff, vj, vhfopt.coeff)
```
Comment (wxj6000) [inline gpu4pyscf/scf/jk.py:163]: Same here.

File: gpu4pyscf/scf/jk.py [hunk 4]
```diff
@@ -0,0 +1,657 @@
...
+        vj = vj + vj.transpose(0,2,1)
```
Comment (wxj6000) [inline gpu4pyscf/scf/jk.py:164]: use transpose_sum

File: gpu4pyscf/scf/jk.py [hunk 5]
```diff
@@ -0,0 +1,657 @@
...
+        omega = mol.omega
+
+    mol = vhfopt.mol
+    nao, nao_orig = vhfopt.coeff.shape
+
+    dm = cp.asarray(dm, order='C')
+    dms = dm.reshape(-1,nao_orig,nao_orig)
+    n_dm = dms.shape[0]
+    assert n_dm == 1
+    dms = cp.einsum('pi,nij,qj->npq', vhfopt.coeff, dms, vhfopt.coeff)
```
Comment (wxj6000) [inline gpu4pyscf/scf/jk.py:191]: replace `cp.einsum`

File: gpu4pyscf/scf/jk.py [hunk 6]
```diff
@@ -0,0 +1,657 @@
...
+            log.debug1('%s wall time %.2f', llll, t)
+        cp.cuda.Stream.null.synchronize()
+        log.timer_debug1('cuda kernel', *t2)
+
+    vj_xyz = vj_xyz.get()
+    vj = np.empty_like(dms)
+    libvhf_rys.transform_xyz_to_cart(
+        vj.ctypes, vj_xyz.ctypes, ao_loc.ctypes, pair_loc.ctypes,
+        mol._bas.ctypes, ctypes.c_int(mol.nbas), mol._env.ctypes)
+    vj = cp.einsum('pi,npq,qj->nij', vhfopt.coeff, cp.asarray(vj), vhfopt.coeff)
```
Comment (wxj6000) [inline gpu4pyscf/scf/jk.py:303]: replace `cp.einsum` with `contract`

File: gpu4pyscf/scf/jk.py [hunk 7]
```diff
@@ -0,0 +1,657 @@
...
+    vj = vj + vj.transpose(0,2,1)
```
Comment (wxj6000) [inline gpu4pyscf/scf/jk.py:304]: transpose_sum

File: gpu4pyscf/lib/gvhf-rys/rys_roots.cu [hunk 8]
```diff
@@ -0,0 +1,72 @@
+#include "rys_roots.cuh"
```
Comment (wxj6000) [inline gpu4pyscf/lib/gvhf-rys/rys_roots.cu:1]: Is this file the same as rys_roots_unrolled.cu?

File: gpu4pyscf/lib/gvhf-rys/unrolled_rys_j.cu [hunk 9]
```diff
@@ -0,0 +1,5499 @@
...
+        double *dm = jk.dm;
+        for (int i_dm = 0; i_dm < jk.n_dm; ++i_dm) {
+            atomicAdd(vj+ij_pair0+0, gout_0_0*dm[kl_pair0+0]);
+            atomicAdd(vj+kl_pair0+0, gout_0_0*dm[ij_pair0+0]);
+            vj += nao_pairs;
+            dm += nao_pairs;
+        }
+    }
+}
+__global__ __maxnreg__(128)
```
Comment (wxj6000) [inline gpu4pyscf/lib/gvhf-rys/unrolled_rys_j.cu:155]: This requires CUDA > 12.4

File: gpu4pyscf/lib/gvhf-rys/gamma_inc.cu [hunk 10]
```diff
@@ -0,0 +1,50 @@
+#include <float.h>
+#define LSUM_MAX        (LMAX*4)
```
Comment (wxj6000) [inline gpu4pyscf/lib/gvhf-rys/gamma_inc.cu:2]: Is gamma_inc.cu the same as gamma_inc_unrolled.cu?

File: gpu4pyscf/lib/gvhf-rys/rys_roots.cu (hunk 8 above)
Comment (sunqm) [reply gpu4pyscf/lib/gvhf-rys/rys_roots.cu:1]: No. different ways to assign threads.

File: gpu4pyscf/lib/gvhf-rys/gamma_inc.cu (hunk 10 above)
Comment (sunqm) [reply gpu4pyscf/lib/gvhf-rys/gamma_inc.cu:2]: No. they are different ways to assign threads.

File: gpu4pyscf/scf/jk.py [hunk 11]
```diff
@@ -0,0 +1,669 @@
...
+    lij = li + lj
+    lkl = lk + ll
+    nmax = max(lij, lkl)
+    nps = l_ctr_pattern[:,1]
+    ij_prims = nps[0] * nps[1]
+    g_size = (lij+1)*(lkl+1)
+    nf3_ij = (lij+1)*(lij+2)*(lij+3)/6
+    nf3_kl = (lkl+1)*(lkl+2)*(lkl+3)/6
+    nroots = (li+lj+lk+ll) // 2 + 1
+    lmax = 4
```
Comment (wxj6000) [inline gpu4pyscf/scf/jk.py:642]: This is hardcoded?

File: gpu4pyscf/scf/jk.py [hunk 12]
```diff
@@ -0,0 +1,669 @@
...
+from gpu4pyscf.lib.cupy_helper import load_library, condense, sandwich_dot, transpose_sum
+from gpu4pyscf.__config__ import props as gpu_specs
+from gpu4pyscf.lib import logger
+
+__all__ = [
+    'get_jk', 'get_j',
+]
+
+PTR_BAS_COORD = 7
+LMAX = 4
```
Comment (wxj6000) [inline gpu4pyscf/scf/jk.py:21]: Are we supporting L > 4?

File: gpu4pyscf/scf/jk.py (hunk 11 above)
Comment (sunqm) [reply gpu4pyscf/scf/jk.py:708]: It's not the angular moment of orbital basis. This is a special/temporary treatment for j-engine.

File: gpu4pyscf/scf/jk.py (hunk 12 above)
Comment (sunqm) [reply gpu4pyscf/scf/jk.py:20]: They are processed on CPU.

**Review by wxj6000 — APPROVED** [review]

**Pull request merged**